### PR TITLE
Add ROSpan StartsWith and EndsWith string-like APIs with StringComparison

### PR DESCRIPTION
**Pull request by ahsonkhan**
Part of https://github.com/dotnet/corefx/issues/21395#issuecomment-359906138

- StartsWith
- EndsWith

Depends on: https://github.com/dotnet/coreclr/pull/16223 and https://github.com/dotnet/coreclr/pull/16418

**TODO:** ~Add tests.~

cc @jkotas, @stephentoub, @KrzysztofCwalina, @tarekgh, @JeremyKuhne

### CODE REVIEW

File: src/System.Memory/src/System/MemoryExtensions.Fast.cs [hunk 1]
```diff
@@ -11,6 +11,24 @@ namespace System
     /// </summary>
     public static partial class MemoryExtensions
     {
+        /// <summary>
+        /// Determines whether the end of the <paramref name="span"/> matches the specified <paramref name="value"/> when compared using the specified <paramref name="comparison"/> option.
+        /// </summary>
+        /// <param name="span">The source span.</param>
+        /// <param name="value">The sequence to compare to the end of the source span.</param>
+        /// <param name="comparison">One of the enumeration values that determines how the <paramref name="span"/> and <paramref name="value"/> are compared.</param>
+        public static bool EndsWith(this ReadOnlySpan<char> span, ReadOnlySpan<char> value, StringComparison comparison)
```
Comment (justinvp) [inline src/System.Memory/src/System/MemoryExtensions.Fast.cs]: `comparison` -> `comparisonType` to match the naming used everywhere else.

File: src/System.Memory/src/System/MemoryExtensions.Portable.cs [hunk 2]
```diff
@@ -12,6 +12,34 @@ namespace System
...
+        /// <summary>
+        /// Determines whether the end of the <paramref name="span"/> matches the specified <paramref name="value"/> when compared using the specified <paramref name="comparison"/> option.
+        /// </summary>
+        /// <param name="span">The source span.</param>
+        /// <param name="value">The sequence to compare to the end of the source span.</param>
+        /// <param name="comparison">One of the enumeration values that determines how the <paramref name="span"/> and <paramref name="value"/> are compared.</param>
+        public static bool EndsWith(this ReadOnlySpan<char> span, ReadOnlySpan<char> value, StringComparison comparison)
+        {
+            string sourceString = span.ToString();
+            string valueString = value.ToString();
```
Comment (stephentoub) [inline src/System.Memory/src/System/MemoryExtensions.Portable.cs]: Ouch; this is really unfortunate.  We should consider only exposing these for core.

If we keep these, we should at least do a reasonable implementation for Ordinal and OrdinalIgnoreCase, and presumably we could do so for Invariant as well?  I agree the CurrentCulture ones will be problematic.

Comment (tarekgh) [reply src/System.Memory/src/System/MemoryExtensions.Portable.cs]: Invariant would be as the current culture (if the string is not fully ASCII). 

I think there is opportunity to optimize for the ordinal and ordinal ignore casing but I think we already saying the portable implementation would be slower than the core one and whatever we do will not be able to match the core performance (i guess).

not exposing these would be an issue for anyone want to write a library that work on net core and down level too. I prefer keeping it and document it as it is slower than the core one and we can try optimize here as much we can.

---
In reply to: [166251624](https://github.com/dotnet/corefx/pull/26880#discussion_r166251624) [](ancestors = 166251624)

Comment (ahsonkhan) [reply src/System.Memory/src/System/MemoryExtensions.Portable.cs]: I can file an issue to optimize this method for the ordinal cases. We do have the generic StartsWith method that take a T, which does not take SringComparison that can be used for one of the two scenarios.

FYI
>     // Those need access to globalization APIs. We'll also expose them from
    // the .NET Framework OOB (slow span). They will try to extract the string
    // from the underlying span (because slow span stores it) -- or -- allocate
    // a new string. This avoids bifurcating the API surface.

Comment (stephentoub) [reply src/System.Memory/src/System/MemoryExtensions.Portable.cs]: > They will try to extract the string from the underlying span (because slow span stores it)

Can this change do this then?

Comment (ahsonkhan) [reply src/System.Memory/src/System/MemoryExtensions.Portable.cs]: I haven't been able to figure out how to do that. The underlying data is stored as a Pinnable<char>, and the globalization APIs work on strings (not char*).

Comment (stephentoub) [reply src/System.Memory/src/System/MemoryExtensions.Portable.cs]: It looks like we're using Unsafe.As to cast the object to a Pinnable.  Can you just do a typecheck on it and cast it back? e.g. `object o = _pinnable; if (o is string s) { ... }`?

Comment (ahsonkhan) [reply src/System.Memory/src/System/MemoryExtensions.Portable.cs]: If that is doable, I can definitely update our ToString implementation and take advantage of that as well.

File: src/System.Memory/src/System/ReadOnlySpan.Portable.cs [hunk 3]
```diff
@@ -193,6 +193,10 @@ public override string ToString()
         {
             if (typeof(T) == typeof(char))
             {
+                if (_pinnable is string str)
```
Comment (jkotas) [inline src/System.Memory/src/System/ReadOnlySpan.Portable.cs]: This is not enough. You have to also check that the offset and length are for the whole string (ie that the Span is not pointing at substring).

Comment (jkotas) [inline src/System.Memory/src/System/ReadOnlySpan.Portable.cs]: This is not enough. You have to also check that the offset and length are for the whole string (ie that the Span is not pointing at substring).

Comment (jkotas) [inline src/System.Memory/src/System/ReadOnlySpan.Portable.cs]: Note that the optimizing compilers can figure out here that `Pinnable<T>` can never be a `string`, and turn this condition into false.

I think you may want to pass it through `Unsafe.As` here to strip the type information here. It is not guaranteed to work, but it is increasing the changes that it will.

Comment (jkotas) [reply src/System.Memory/src/System/ReadOnlySpan.Portable.cs]: We may want to have a slow-span specific test for this so that we will know when it does not work.

Comment (stephentoub) [reply src/System.Memory/src/System/ReadOnlySpan.Portable.cs]: > Note that the optimizing compilers can figure out here that Pinnable<T> can never be a string

Even the C# compiler will... that's why in my example I stored it into an object first, otherwise the C# compiler just emit the cast as a hard-coded false. Though as you say an even better compiler could see through the object cast as well.

Comment (ahsonkhan) [reply src/System.Memory/src/System/ReadOnlySpan.Portable.cs]: Would adding offset and length checks be sufficient here?

```C#
if (obj is string str && _byteOffset == 0 && _length == str.Length)
```

File: src/System.Memory/src/System/Span.Portable.cs [hunk 4]
```diff
@@ -305,6 +305,10 @@ public override string ToString()
         {
             if (typeof(T) == typeof(char))
             {
+                if (_pinnable is string str)
```
Comment (ahsonkhan) [inline src/System.Memory/src/System/Span.Portable.cs]: Can this ever be true for Span or should I remove this check?

File: src/System.Memory/src/System/ReadOnlySpan.Portable.cs (hunk 3 above)
Comment (jkotas) [reply src/System.Memory/src/System/ReadOnlySpan.Portable.cs]: You need to use `StringAdjustment` for the offset.

File: src/System.Memory/src/System/MemoryExtensions.Portable.cs (hunk 2 above)
Comment (ahsonkhan) [reply src/System.Memory/src/System/MemoryExtensions.Portable.cs]: Is there a way to do something similar for the FastSpan ToString which has a ByReference field?

**Comment by ahsonkhan** [comment]
For slow span, if we get a read/write Span from Memory, we have Pinnable\<T\> != null and hence ToString() can return the underlying string.

But, if we get construct a span from void* and length, Pinnable\<T\> == null and then ToString() returns a new string.

```C#
string original = TestHelpers.BuildString(10, 42);
ReadOnlyMemory<char> readOnlyMemory = original.AsReadOnlyMemory();
Memory<char> memory = MemoryMarshal.AsMemory(readOnlyMemory);

fixed (char* pOriginal = original)
fixed (char* pString1 = memory.Span.ToString())
{
    Assert.Equal((int)pOriginal, (int)pString1); // passes the test
}

fixed (char* pOriginal = original)
{
    Span<char> span = new Span<char>(pOriginal, original.Length);
    fixed (char* pString1 = span.ToString())
    {
        Assert.Equal((int)pOriginal, (int)pString1); // fails the test
    }
}
```

File: src/System.Memory/src/System/MemoryExtensions.Portable.cs (hunk 2 above)
Comment (jkotas) [reply src/System.Memory/src/System/MemoryExtensions.Portable.cs]: No, there is not a reasonable way to do that for fast span.

File: src/System.Memory/tests/ReadOnlySpan/ToString.cs [hunk 5]
```diff
@@ -43,11 +43,53 @@ public static void ToStringChar_Empty()
...
+            ReadOnlySpan<char> span = original.AsReadOnlySpan();
+
+            string returnedString = span.ToString();
+            string returnedStringUsingSlice = span.Slice(0, original.Length).ToString();
+
+            string subString1 = span.Slice(1).ToString();
+            string subString2 = span.Slice(0, 2).ToString();
+            string subString3 = span.Slice(1, 2).ToString();
+
+            Assert.Equal("RDDNEGSNET", returnedString);
```
Comment (joshfree) [inline src/System.Memory/tests/ReadOnlySpan/ToString.cs]: nit: shouldn't these use `original` instead of the hardcoded literal

**Comment by ahsonkhan** [comment]
@dotnet-bot test this please

**Comment by ahsonkhan** [comment]
@dotnet-bot test this please

File: src/System.Memory/src/System/ReadOnlySpan.Portable.cs [hunk 6]
```diff
@@ -193,6 +193,11 @@ public override string ToString()
         {
             if (typeof(T) == typeof(char))
             {
+                object obj = Unsafe.As<Pinnable<T>, object>(ref Unsafe.AsRef(_pinnable));
+                if (obj is string str && _byteOffset == MemoryExtensions.StringAdjustment && _length == str.Length)
```
Comment (stephentoub) [inline src/System.Memory/src/System/ReadOnlySpan.Portable.cs]: As in https://github.com/dotnet/corefx/pull/27170/, how about doing the _byteOffset check before doing any of the casting / type checking?

File: src/System.Memory/tests/ReadOnlySpan/ToString.cs [hunk 7]
```diff
@@ -43,11 +43,56 @@ public static void ToStringChar_Empty()
...
+            Assert.Equal(original.Substring(0, 2), subString2);
+            Assert.Equal(original.Substring(1, 2), subString3);
+
+            fixed (char* pOriginal = original)
+            fixed (char* pString1 = returnedString)
+            fixed (char* pString2 = returnedStringUsingSlice)
+            {
+                Assert.Equal((int)pOriginal, (int)pString1);
+                Assert.Equal((int)pOriginal, (int)pString2);
+            }
```
Comment (stephentoub) [inline src/System.Memory/tests/ReadOnlySpan/ToString.cs]: Why not just use Assert.Same?

File: src/System.Memory/tests/ReadOnlySpan/ToString.cs [hunk 8]
```diff
@@ -43,11 +43,56 @@ public static void ToStringChar_Empty()
...
+            fixed (char* pSubString2 = subString2)
+            fixed (char* pSubString3 = subString3)
+            {
+                Assert.NotEqual((int)pOriginal, (int)pSubString1);
+                Assert.NotEqual((int)pOriginal, (int)pSubString2);
+                Assert.NotEqual((int)pOriginal, (int)pSubString3);
+                Assert.NotEqual((int)pSubString1, (int)pSubString2);
+                Assert.NotEqual((int)pSubString1, (int)pSubString3);
+                Assert.NotEqual((int)pSubString2, (int)pSubString3);
+            }
```
Comment (stephentoub) [inline src/System.Memory/tests/ReadOnlySpan/ToString.cs]: And Assert.NotSame?

File: src/System.Memory/tests/ReadOnlySpan/ToString.cs [hunk 9]
```diff
@@ -43,11 +43,56 @@ public static void ToStringChar_Empty()
...
         {
             string[] a = { "a", "b", "c" };
             var span = new ReadOnlySpan<string>(a);
             Assert.Equal("System.ReadOnlySpan<String>[3]", span.ToString());
         }
+
+        // This test is only relevant for portable span
+#if FEATURE_PORTABLE_SPAN
+        [Fact]
+#endif
```
Comment (stephentoub) [inline src/System.Memory/tests/ReadOnlySpan/ToString.cs]: Is this the convention used in this project?  Elsewhere in corefx we've used SkipOnTargetFramework, e.g. https://github.com/dotnet/corefx/pull/27170/files#diff-f8d6f156088e92893813570d9608526eR63

File: src/System.Memory/src/System/ReadOnlySpan.Portable.cs [hunk 10]
```diff
@@ -193,6 +193,11 @@ public override string ToString()
         {
             if (typeof(T) == typeof(char))
             {
+                object obj = Unsafe.As<Pinnable<T>, object>(ref Unsafe.AsRef(_pinnable));
```
Comment (stephentoub) [inline src/System.Memory/src/System/ReadOnlySpan.Portable.cs]: Can this just be:
```C#
object obj = Unsafe.As<object>(_pinnable);
```
?

Comment (stephentoub) [reply src/System.Memory/src/System/ReadOnlySpan.Portable.cs]: Nit: (Also, a comment indicating why this is used rather than just `if (_pinnable is string str ...)` would be helpful.)

File: src/System.Memory/src/System/Span.Portable.cs [hunk 11]
```diff
@@ -305,6 +305,11 @@ public override string ToString()
         {
             if (typeof(T) == typeof(char))
             {
+                object obj = Unsafe.As<Pinnable<T>, object>(ref Unsafe.AsRef(_pinnable));
+                if (obj is string str && _byteOffset == MemoryExtensions.StringAdjustment && _length == str.Length)
+                {
+                    return str;
+                }
```
Comment (stephentoub) [inline src/System.Memory/src/System/Span.Portable.cs]: Is this change needed?  When will a `Span<char>` (rather than a `ReadOnlySpan<char>`) wrap a `string`?

Comment (ahsonkhan) [reply src/System.Memory/src/System/Span.Portable.cs]: See code sample here: https://github.com/dotnet/corefx/pull/26880#issuecomment-363632190

File: src/System.Memory/src/System/ReadOnlySpan.Portable.cs (hunk 10 above)
Comment (ahsonkhan) [reply src/System.Memory/src/System/ReadOnlySpan.Portable.cs]: I will merge the changes from your PR into here which will include the comments, test updates, etc.

Thanks!

File: src/System.Memory/src/System/ReadOnlySpan.Portable.cs (hunk 6 above)
Comment (ahsonkhan) [reply src/System.Memory/src/System/ReadOnlySpan.Portable.cs]: Will do.

File: src/System.Memory/src/System/Span.Portable.cs (hunk 11 above)
Comment (stephentoub) [reply src/System.Memory/src/System/Span.Portable.cs]: I personally don't think it's worthwhile to add this check to `Span<char>`.  It's such a corner case to optimize for, and at the expense of adding a type check to other `Span<char>.ToString()` calls.  And in general folks can't depend on ToString producing the same string from multiple calls, so I don't consider the discrepancy a real issue.

File: src/System.Memory/src/System/ReadOnlySpan.Portable.cs (hunk 10 above)
Comment (stephentoub) [reply src/System.Memory/src/System/ReadOnlySpan.Portable.cs]: Cool, thanks.  Didn't realize you'd already done the initial work (I remember we'd talked about it, but then couldn't find it in the repo :smile:)

File: src/System.Memory/tests/ReadOnlySpan/ToString.cs (hunk 9 above)
Comment (ahsonkhan) [reply src/System.Memory/tests/ReadOnlySpan/ToString.cs]: I am not sure if NetFramework is the only one TargetFrameworkMoniker where we would skip this test.

Comment (stephentoub) [reply src/System.Memory/tests/ReadOnlySpan/ToString.cs]: Ok.

Comment (ahsonkhan) [reply src/System.Memory/tests/ReadOnlySpan/ToString.cs]: But given [this](https://github.com/dotnet/corefx/blob/7059f8b316723f5db64e4b1d28e2b1e8b137ab1e/Documentation/project-docs/developer-guide.md#skipontargetframeworkattribute), I will change it:

File: src/System.Memory/src/System/Span.Portable.cs (hunk 11 above)
Comment (ahsonkhan) [reply src/System.Memory/src/System/Span.Portable.cs]: I don't have a strong opinion on this. I am fine with removing this check for Span. @KrzysztofCwalina, @jkotas - any objections?

**Comment by ahsonkhan** [comment]
@dotnet-bot test Linux x64 Release Build

**Comment by ahsonkhan** [comment]
@dotnet-bot test this please
